File: frontend/src/lib/types/analytics.ts
// ---------------------------------------------------------------------------
// Analytics redesign — new shape (SPEC-analytics-redesign.md §3).
//
// Mirrors backend/app/schemas/analytics.py AnalyticsResponse. Legacy types
// (FunnelData, HeroCounters, ResumePerformanceData, ResponseTimeData,
// AIOperationCount, DataSufficiency, and the legacy AnalyticsResponse fields)
// are retained at the bottom marked @deprecated until Phase 10.5 cleanup so
// that DEMO_ANALYTICS_DATA and the orphaned legacy card components still
// type-check while the new page lands.
// ---------------------------------------------------------------------------

export type AnalyticsPeriod = "7d" | "30d" | "90d" | "all";

/** One bucket of the applications timeline (day/week/month — granularity is
 * server-determined; date is always the start of the bucket as ISO date). */
export interface TimelineBucket {
  date: string;
  count: number;
}

/** One day in the activity heatmap. Always daily granularity and covers the
 * last ~91 days regardless of the page period selector. */
export interface DailyActivity {
  date: string;
  count: number;
}

/** Best 7-day window of applications, all-time. */
export interface PersonalBest {
  start_date: string;
  end_date: string;
  count: number;
  metric: "applications";
}

/** One headline KPI card on the strip. */
export interface KpiMetric {
  current: number;
  delta: number;
  sparkline: number[];
}

/** One stage row of the cumulative funnel. `percentage_of_top` is count /
 * saved-stage-count (0 when saved == 0). */
export interface FunnelStage {
  stage: "saved" | "applied" | "screening" | "interview" | "offer";
  count: number;
  percentage_of_top: number;
}

/** One slice of the active-pipeline donut. `status` is an ApplicationStatus
 * enum value from the backend (kept as string here — the donut component
 * narrows when it maps colors). */
export interface StatusCount {
  status: string;
  count: number;
}

/** One bucket of the document-generation timeline. Granularity matches the
 * applications timeline so the charts line up visually. */
export interface GenerationBucket {
  date: string;
  resumes: number;
  cover_letters: number;
}

export type ActivityEventType =
  | "status_change_offer"
  | "status_change_interview"
  | "status_change_screening"
  | "status_change_applied"
  | "status_change_rejected"
  | "resume_tailored"
  | "cover_letter_generated"
  | "job_saved";

/** One row in the recent-activity feed. `id` is composite
 * (`{source}_{source_id}`) so list keys never collide across the three
 * event sources.
 *
 * The new structured fields (`company`, `role`, `to_status`) are the canonical
 * source of truth — the title is now built client-side via `useTranslations`
 * off `type` + `company` + `role`. The legacy `title` and `meta` fields are
 * still returned by the backend (Russian-hardcoded) but should not be read. */
export interface ActivityEvent {
  id: string;
  type: ActivityEventType;
  occurred_at: string;
  /** @deprecated Russian-hardcoded backend pre-format. Build the title
   * client-side from `type` + `company` + `role` using `useTranslations`. */
  title: string;
  /** @deprecated Russian-hardcoded backend meta (e.g. "новый статус: ..."),
   * redundant with the localised action verb. Do not render. */
  meta: string | null;
  application_id: string | null;
  /** Company name when the event has an associated application/job (e.g. "Acme"). */
  company: string | null;
  /** Role/title when the event has an associated application/job (e.g. "Senior Frontend Engineer"). */
  role: string | null;
  /** Destination status for `status_change_*` events; null for other event types. */
  to_status: string | null;
}

/** Full payload returned by `GET /api/v1/analytics?period={period}`. */
export interface AnalyticsResponse {
  period: AnalyticsPeriod;
  period_start: string;
  period_end: string;

  // Hero block — big number + dual-line chart.
  applications_current: number;
  applications_previous: number;
  applications_delta: number;
  applications_timeline: TimelineBucket[];
  applications_timeline_previous: TimelineBucket[];

  // Streak panel — heatmap is period-independent (always last 91 days).
  current_streak_days: number;
  longest_streak_days: number;
  daily_activity: DailyActivity[];
  personal_best: PersonalBest | null;

  // KPI strip — 4 cards with sparklines.
  jobs_saved: KpiMetric;
  applied: KpiMetric;
  interviews: KpiMetric;
  offers: KpiMetric;

  // Funnel + insight callout.
  funnel: FunnelStage[];
  funnel_insight: string | null;

  // Active-pipeline donut.
  status_breakdown: StatusCount[];

  // Generations area chart.
  generations_total_resumes: number;
  generations_total_cover_letters: number;
  generations_timeline: GenerationBucket[];

  // Recent-activity feed.
  recent_activity: ActivityEvent[];
}

// ---------------------------------------------------------------------------
// LEGACY shape — kept until Phase 10.5 cleanup so that DEMO_ANALYTICS_DATA
// and the orphaned legacy card components (HeroStats, ConversionFunnel,
// ApplicationsTimeline, StatusBreakdown, ResumePerformance, ResponseTime,
// AIOperationsBreakdown, InsufficientDataState) still type-check.
//
// Do not extend or add new usages — these interfaces will be deleted.
// ---------------------------------------------------------------------------

/** @deprecated Legacy funnel stage with drop-off semantics. Use FunnelStage above. */
export interface FunnelStageLegacy {
  stage: string;
  count: number;
  pct_of_saved: number;
  drop_off_pct: number | null;
}

/** @deprecated Legacy funnel wrapper. */
export interface FunnelData {
  stages: FunnelStageLegacy[];
  insight: string | null;
}

/** @deprecated Legacy month-granularity bucket. Use TimelineBucket above. */
export interface TimelineMonth {
  month: string;
  count: number;
}

/** @deprecated Legacy resume sparkline point. */
export interface ResumeSparklinePoint {
  score: number;
  resume_id: string;
  resume_name: string;
}

/** @deprecated Legacy resume-performance block. */
export interface ResumePerformanceData {
  avg_score: number | null;
  sparkline: ResumeSparklinePoint[];
  best_resume_id: string | null;
  best_resume_name: string | null;
  best_score: number | null;
  worst_resume_id: string | null;
  worst_resume_name: string | null;
  worst_score: number | null;
  trend: "improving" | "declining" | "stable" | null;
}

/** @deprecated Legacy AI operation count. */
export interface AIOperationCount {
  operation_type: string;
  count: number;
}

/** @deprecated Legacy hero counters block. Use KpiMetric per-card above. */
export interface HeroCounters {
  jobs_saved: number;
  applied: number;
  interviews: number;
  offers: number;
  jobs_saved_delta: number;
  applied_delta: number;
  interviews_delta: number;
  offers_delta: number;
}

/** @deprecated Legacy response-time block (removed from the redesigned page). */
export interface ResponseTimeData {
  avg_days: number | null;
}

/** @deprecated Legacy data-sufficiency gate (removed from the redesigned page). */
export interface DataSufficiency {
  total_jobs: number;
  total_applications: number;
  total_scored_resumes: number;
  threshold_jobs: number;
  threshold_applications: number;
  threshold_scored_resumes: number;
  has_sufficient_data: boolean;
}

/** @deprecated Legacy analytics response shape. Use AnalyticsResponse above.
 * Kept so DEMO_ANALYTICS_DATA and the orphaned card components compile until
 * Phase 10.5 cleanup. */
export interface AnalyticsResponseLegacy {
  funnel: FunnelData;
  timeline: TimelineMonth[];
  status_breakdown: StatusCount[];
  resume_performance: ResumePerformanceData;
  ai_operations: AIOperationCount[];
  response_time: ResponseTimeData;
  hero_counters: HeroCounters;
  data_sufficiency: DataSufficiency;
}

File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel

from app.models.enums import ApplicationStatus

# ---------------------------------------------------------------------------
# Analytics redesign — new shape (SPEC-analytics-redesign.md §3.1, §3.2)
#
# These models replace the legacy AnalyticsResponse / FunnelData /
# ResumePerformanceData / HeroCounters / etc. shape. The legacy classes
# below this section are intentionally retained until Phase 10.5 cleanup
# so that any in-flight imports keep working while Phases 2–3 land.
# ---------------------------------------------------------------------------


Period = Literal["7d", "30d", "90d", "all"]


class TimelineBucket(BaseModel):
    """One bucket of the applications timeline.

    Granularity (day / week / month) is determined by the requesting
    period — the bucket date is the start of the bucket regardless.
    """

    date: date
    count: int


class DailyActivity(BaseModel):
    """One day in the activity heatmap (always daily granularity).

    Period-independent: the heatmap always covers the last ~91 days so
    the streak and personal-best computations stay stable when the user
    switches the page period selector.
    """

    date: date
    count: int


class PersonalBest(BaseModel):
    """Best 7-day window of applications, all-time."""

    start_date: date
    end_date: date
    count: int
    metric: Literal["applications"]


class KpiMetric(BaseModel):
    """One headline KPI on the strip (saved / applied / interview / offer)."""

    current: int
    delta: int
    sparkline: list[int]


class FunnelStage(BaseModel):
    """One stage row of the cumulative funnel.

    Cumulative semantics: e.g. ``interview.count`` includes every
    application that ever reached interview, even if it has since
    moved to offer or rejected. ``percentage_of_top`` is the count
    divided by the ``saved`` stage count (or 0 if saved == 0).
    """

    stage: Literal["saved", "applied", "screening", "interview", "offer"]
    count: int
    percentage_of_top: float


class StatusCount(BaseModel):
    """One slice of the active-pipeline donut."""

    status: ApplicationStatus
    count: int


class GenerationBucket(BaseModel):
    """One bucket of the document-generation timeline.

    Granularity matches the applications timeline for visual alignment.
    Counts are split by operation type so the chart can stack them.
    """

    date: date
    resumes: int
    cover_letters: int


ActivityEventType = Literal[
    "status_change_offer",
    "status_change_interview",
    "status_change_screening",
    "status_change_applied",
    "status_change_rejected",
    "resume_tailored",
    "cover_letter_generated",
    "job_saved",
]


class ActivityEvent(BaseModel):
    """One row in the recent-activity feed.

    ``id`` is composite (``f"{source}_{source_id}"``) so the frontend
    can key a list without colliding rows from the three event sources.
    ``application_id`` is set whenever the event has a parent
    application (i.e. always except for purely synthetic future events).

    Localization note: ``title`` and ``meta`` are DEPRECATED — the
    backend used to do string formatting + localization (Russian), which
    is the wrong layer. They are still populated for backward compat
    during the frontend migration; once the frontend renders from the
    structured fields below, the legacy fields can be dropped.

    The structured fields (``company``, ``role``, ``to_status``) let the
    frontend build localized labels from translation strings. They are
    populated for every event type where a parent application can be
    resolved.
    """

    id: str
    type: ActivityEventType
    occurred_at: datetime
    title: str  # DEPRECATED — populated for transition; frontend builds from structured fields below
    meta: str | None = None  # DEPRECATED for status changes; kept for future use (match score, tone, etc.)
    application_id: str | None = None

    # New structured fields — populate for relevant event types.
    company: str | None = None   # populated when application_id is known
    role: str | None = None      # populated when application_id is known
    to_status: str | None = None # populated for status_change_* events (value like "applied", "interview", "offer")


class AnalyticsResponse(BaseModel):
    """Full payload for ``GET /api/analytics?period={...}``.

    Single round-trip: every block on /dashboard/analytics is fed from
    one of the fields below. Computation is orchestrated by
    AnalyticsService.get_full_analytics in Phase 2; caching is handled
    in Phase 3.
    """

    period: Period
    period_start: date
    period_end: date

    # Hero block — big number + dual-line chart.
    applications_current: int
    applications_previous: int
    applications_delta: int
    applications_timeline: list[TimelineBucket]
    applications_timeline_previous: list[TimelineBucket]

    # Streak panel — heatmap is period-independent (always last 91 days).
    current_streak_days: int
    longest_streak_days: int
    daily_activity: list[DailyActivity]
    personal_best: PersonalBest | None = None

    # KPI strip — 4 cards with sparklines.
    jobs_saved: KpiMetric
    applied: KpiMetric
    interviews: KpiMetric
    offers: KpiMetric

    # Funnel + insight callout.
    funnel: list[FunnelStage]
    funnel_insight: str | None = None

    # Active-pipeline donut.
    status_breakdown: list[StatusCount]

    # Generations area chart.
    generations_total_resumes: int
    generations_total_cover_letters: int
    generations_timeline: list[GenerationBucket]

    # Recent-activity feed.
    recent_activity: list[ActivityEvent]


# ---------------------------------------------------------------------------
# LEGACY shape — kept until Phase 10.5 cleanup so any callers still
# importing these names don't break mid-migration. Do not add new
# usages; do not extend these models. New analytics features go on
# AnalyticsResponse above.
# ---------------------------------------------------------------------------


class FunnelStageLegacy(BaseModel):
    """LEGACY funnel stage shape (drop_off_pct semantics).

    Used only by analytics_service.compute_funnel and
    analytics_insights.get_funnel_insight pending Phase 2 rewrite.
    Do not use in new code — see ``FunnelStage`` above for the
    cumulative-counts shape that lands with the redesign.
    """

    stage: str
    count: int
    pct_of_saved: float
    drop_off_pct: float | None  # vs previous stage


class FunnelData(BaseModel):
    stages: list[FunnelStageLegacy]
    insight: str | None  # filled by insight generator


class TimelineMonth(BaseModel):
    month: str  # "2025-10"
    count: int


class ResumeSparklinePoint(BaseModel):
    score: int
    resume_id: str
    resume_name: str


class ResumePerformanceData(BaseModel):
    avg_score: float | None
    sparkline: list[ResumeSparklinePoint]  # last 10 tailored resumes
    best_resume_id: str | None
    best_resume_name: str | None
    best_score: int | None
    worst_resume_id: str | None
    worst_resume_name: str | None
    worst_score: int | None
    trend: str | None  # "improving" | "declining" | "stable" — filled by insight gen


class AIOperationCount(BaseModel):
    operation_type: str
    count: int


class HeroCounters(BaseModel):
    jobs_saved: int
    applied: int
    interviews: int
    offers: int
    jobs_saved_delta: int
    applied_delta: int
    interviews_delta: int
    offers_delta: int


class ResponseTimeData(BaseModel):
    avg_days: float | None


class DataSufficiency(BaseModel):
    total_jobs: int
    total_applications: int
    total_scored_resumes: int
    threshold_jobs: int       # 5
    threshold_applications: int  # 5
    threshold_scored_resumes: int  # 3
    has_sufficient_data: bool  # true when total_applications >= 5


class AnalyticsResponseLegacy(BaseModel):
    """LEGACY full analytics payload.

    Still served by the existing /api/v1/analytics endpoint until
    Phase 3 of the redesign rewires the router to assemble the new
    ``AnalyticsResponse`` shape. Kept under a distinct name so the new
    response model can claim the canonical ``AnalyticsResponse`` symbol
    while the legacy code path keeps compiling.
    """

    funnel: FunnelData
    timeline: list[TimelineMonth]  # 6 months
    status_breakdown: list[StatusCount]
    resume_performance: ResumePerformanceData
    ai_operations: list[AIOperationCount]
    response_time: ResponseTimeData
    hero_counters: HeroCounters
    data_sufficiency: DataSufficiency
